Added support for a maximum content length property. The property rep…

eschlenz opened the pull request:
…resents a threshold for content that should be saved verbatim, or not.

@jgilfelt 

**WHAT:** This PR introduces a "max content length" property to the `ChuckInterceptor`. A request or response body that exceeds the new property's value will **not** be recorded. That is, there will still be a record of the transaction, but the content itself is not kept. 

**WHY:** Chuck was crashing on me when I would try to go to the Chuck UI. I determined that the saved response body for a few of my requests were simply too large for SQLite to handle. In such scenarios, I still want to see the transaction in the log. I don't need to see the content however. It made sense to apply this new max content length property to requests as well.

The max content length property is essentially disabled by default (set to Integer.MAX_VALUE). The user can optionally set a max length that works for them.

**MISC:** I tried to stick with the existing code conventions. Thanks for putting this tool together. It's very handy.

joaocsousa commented on library/src/main/java/com/readystatesoftware/chuck/ChuckInterceptor.java:
```diff
@@ -52,6 +52,7 @@
     private Context context;
     private NotificationHelper notificationHelper;
     private boolean showNotification;
+    private int maxContentLength = Integer.MAX_VALUE;
```
If this field will be compared against a Long (```transaction.getRequestContentLength```), shouldn't this field also be a long with default value Long.MAX_VALUE?

eschlenz replied:
Good catch! I'll fix that right now.

joaocsousa commented on library/src/main/java/com/readystatesoftware/chuck/ChuckInterceptor.java:
```diff
@@ -140,26 +159,35 @@ public ChuckInterceptor showNotification(boolean show) {
+                    }
+                }
+                if (isPlaintext(buffer)) {
+                    transaction.setResponseBody(buffer.clone().readString(charset));
```
I'm not sure if this would work, but instead of ignoring the response all together (due to it being too big), couldn't you try to truncate instead to the bytes specified by ```maxContentLength```, via ```buffer.clone().readString(maxContentLength, charset)```?

eschlenz replied:
I thought about doing that, but that would essentially create malformed JSON. So when the FormatUtils.formatJson() runs to try to "pretty up" the JSON for display, it's probably going to barf.

jgilfelt replied:
I agree with @joaocsousa that truncating the response is a better approach. The formatting methods [shouldn't](https://github.com/jgilfelt/chuck/blob/master/library/src/main/java/com/readystatesoftware/chuck/internal/support/FormatUtils.java#L68) [barf](https://github.com/jgilfelt/chuck/blob/master/library/src/main/java/com/readystatesoftware/chuck/internal/support/FormatUtils.java#L82), they'll just return the original text if there are parsing errors, and it would greatly simplify this change.

If we created a method like this:

```java
private String readFromBuffer(Buffer buffer, Charset charset) {
    long bufferSize = buffer.size();
    long maxBytes = Math.min(bufferSize, maxContentLength);
    String body = buffer.readString(maxBytes, charset);
    if (bufferSize > maxContentLength) {
        body += "\n\n--- Content truncated ---";
    }
    return body;
}
```

then it could simply be called where we set the request and response body on the transaction, and there's no need to store these `requestBodyIsTooBig` and `responseBodyIsTooBig` fields in the database. Everything else would work as is.

jgilfelt commented:
Thanks! I've finally got my mock web server setup to send responses large enough to break things, so I will look at this as soon as possible.

I looks like it is Android's CursorWindow that breaks first, whose [hard limit is 2048 KB](https://github.com/android/platform_frameworks_base/blob/master/core/res/res/values/config.xml#L1586).

eschlenz commented:
@joaocsousa @jgilfelt Thanks for the review. I've made the switch from Integer to Long per joaocsuousa's recommendation. I've responded to the other callout (whether or not to keep the partial result).

eschlenz commented:
Fixed the merge conflicts as well.

jgilfelt commented on library/src/main/java/com/readystatesoftware/chuck/ChuckInterceptor.java:
```diff
@@ -74,6 +74,7 @@
     private NotificationHelper notificationHelper;
     private RetentionManager retentionManager;
     private boolean showNotification;
+    private long maxContentLength = Long.MAX_VALUE;
```
I think we should set a conservative default value that will ensure we don't exceed the platform limits, and let users increase it if they need to. We know the CursorWindow limit is 2 MB, but even at that size, I found that a response may still cause ANR issues when displayed with TextView, and potentially consume the application's Java heap space. 

I think `250000` is a good default that will be suitable for most apps, and ensures things wont break.

eschlenz replied:
👍

jgilfelt commented on library/src/main/java/com/readystatesoftware/chuck/ChuckInterceptor.java:
```diff
@@ -97,6 +98,19 @@ public ChuckInterceptor showNotification(boolean show) {
+     * @param max the max length for request/response content.
+     * @return The {@link ChuckInterceptor} instance.
+     */
+    public ChuckInterceptor maxContentLength(long max) {
```
You'll need to shadow any methods that are public API in the [corresponding class](https://github.com/jgilfelt/chuck/blob/master/library-no-op/src/main/java/com/readystatesoftware/chuck/ChuckInterceptor.java) within the `library-no-op` module.

eschlenz replied:
👍

eschlenz commented:
@jgilfelt Saw your change requests. I'll knock that out tomorrow and hit you back.

eschlenz commented:
@jgilfelt @joaocsousa 

I have addressed the feedback you guys gave me:

1. I have shadowed the maxContentLength function in the no-op library.
2. I have removed the "content too big" properties that I had previously added to the HttpTransaction model.
3. The content is now retained, but truncated if it is too long.
4. I have set the default maxContentLength to a more reasonable (and logical) value of 250000L.

With maxContentLength left at its default value:

![screen shot 2017-02-14 at 9 14 55 am](https://cloud.githubusercontent.com/assets/2137515/22938100/fd900dc4-f297-11e6-992e-e18616cf511f.png)

![screen shot 2017-02-14 at 9 14 58 am](https://cloud.githubusercontent.com/assets/2137515/22938110/0271f956-f298-11e6-8b61-401766060914.png)

With maxContentLength set to 5 (for testing):

![screen shot 2017-02-14 at 9 15 31 am](https://cloud.githubusercontent.com/assets/2137515/22938165/267877d0-f298-11e6-9cc5-7c95b9306c4d.png)

![screen shot 2017-02-14 at 9 15 36 am](https://cloud.githubusercontent.com/assets/2137515/22938168/2d70751a-f298-11e6-9f72-102d10714a82.png)
